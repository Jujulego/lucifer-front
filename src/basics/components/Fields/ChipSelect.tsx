import React from 'react';

import {
  Chip, MenuItem,
  FormControl, InputLabel, Select, FormHelperText,
  FormControlProps, SelectProps, ChipProps,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

// Types
interface ChipSelectBaseProps {
  label?: string, helperText?: string,
  options: string[], value?: string[],
  onChange: SelectProps['onChange'],

  ChipProps?: Omit<ChipProps, 'label'>,
  SelectProps?: Omit<SelectProps, 'multiple' | 'value' | 'onChange' | 'renderValue'>
}

export type ChipSelectProps =
  Omit<FormControlProps, keyof ChipSelectBaseProps> &
  ChipSelectBaseProps

// Styles
const useStyles = makeStyles({
  chips: {
    display: 'flex',
    flexWrap: 'wrap'
  },
  chip: {
    marginLeft: 2,
    marginRight: 2
  }
});

// Component
const ChipSelect = (props: ChipSelectProps) => {
  // Props
  const {
    label, helperText,
    options, value = [], onChange,

    ChipProps, SelectProps,

    ...control
  } = props;

  // Render
  const styles = useStyles();

  return (
    <FormControl {...control}>
      { label && <InputLabel>{ label }</InputLabel> }
      <Select {...SelectProps}
        multiple
        value={value} onChange={onChange}
        renderValue={selected => (
          <div className={styles.chips}>
            { (selected as string[]).map(value => (
              <Chip {...ChipProps} className={styles.chip} key={value} label={value} />
            )) }
          </div>
        )}
      >
        { options.map(opt => (
          <MenuItem key={opt} value={opt}>
            { opt }
          </MenuItem>
        )) }
      </Select>
      { helperText && <FormHelperText>{ helperText }</FormHelperText> }
    </FormControl>
  );
};

export default ChipSelect;
