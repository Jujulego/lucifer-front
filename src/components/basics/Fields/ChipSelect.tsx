import React from 'react';

import {
  Chip, MenuItem,
  FormControl, InputLabel, Select, FormHelperText,
  FormControlProps, SelectProps, ChipProps,
} from '@material-ui/core';

// Types
interface ChipSelectBaseProps {
  label?: string, helperText?: string,
  options: string[], value: string[],
  onChange: SelectProps['onChange'],

  ChipProps?: Omit<ChipProps, 'label'>,
  SelectProps?: Omit<SelectProps, 'multiple' | 'value' | 'onChange' | 'renderValue'>
}

export type ChipSelectProps =
  Omit<FormControlProps, keyof ChipSelectBaseProps> &
  ChipSelectBaseProps

// Component
const ChipSelect = (props: ChipSelectProps) => {
  // Props
  const {
    label, helperText,
    options, value, onChange,

    ChipProps, SelectProps,

    ...control
  } = props;

  // Render
  return (
    <FormControl {...control}>
      { label && <InputLabel>{ label }</InputLabel> }
      <Select {...SelectProps}
        multiple
        value={value} onChange={onChange}
        renderValue={selected => (
          <div>
            { (selected as string[]).map(value => (
              <Chip {...ChipProps} key={value} label={value} />
            )) }
          </div>
        )}
      >
        { options.map(opt => (
          <MenuItem key={opt} value={opt}>
            { opt }
          </MenuItem>
        )) }
      </Select>
      { helperText && <FormHelperText>{ helperText }</FormHelperText> }
    </FormControl>
  );
};

export default ChipSelect;