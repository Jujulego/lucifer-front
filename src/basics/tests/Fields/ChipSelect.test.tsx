import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';

import ChipSelect from '../../components/Fields/ChipSelect';

// Setup
let mount: ReturnType<typeof createMount>;
let shallow: ReturnType<typeof createShallow>;

beforeAll(() => {
  mount = createMount();
  shallow = createShallow();
});

afterAll(() => {
  mount.cleanUp();
});

// Tests
it('should render correctly', () => {
  // Render
  const wrapper = shallow(
    <ChipSelect
      label="Test" helperText="Lorem ipsum dolor"
      value={['Test #2']} options={['Test #1', 'Test #2', 'Test #3']}
      onChange={() => {}}
    />
  );

  // Check elements
  expect(wrapper).toMatchSnapshot();
});

it('should change on option click', () => {
  const spy = jest.fn();

  // Render
  const wrapper = mount(
    <ChipSelect
      label="Test" helperText="Lorem ipsum dolor"
      SelectProps={{ open: true }}
      value={['Test #2']} options={['Test #1', 'Test #2', 'Test #3']}
      onChange={spy}
    />
  );

  // Get first option
  const option = wrapper.find('ul > *:first-child li');
  expect(option).toHaveLength(1);

  // Test event
  option.simulate('click');

  expect(spy).toHaveBeenCalledTimes(1);
  expect(spy).toHaveBeenCalledWith(
    expect.objectContaining({
      target: expect.objectContaining({
        value: ['Test #2', 'Test #1']
      })
    }),
    expect.anything()
  )
});
