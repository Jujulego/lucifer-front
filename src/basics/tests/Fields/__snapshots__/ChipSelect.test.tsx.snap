// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<WithStyles(ForwardRef(FormControl))>
  <WithStyles(ForwardRef(InputLabel))>
    Test
  </WithStyles(ForwardRef(InputLabel))>
  <WithStyles(ForwardRef(Select))
    multiple={true}
    onChange={[Function]}
    renderValue={[Function]}
    value={
      Array [
        "Test #2",
      ]
    }
  >
    <WithStyles(ForwardRef(MenuItem))
      key="Test #1"
      value="Test #1"
    >
      Test #1
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="Test #2"
      value="Test #2"
    >
      Test #2
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      key="Test #3"
      value="Test #3"
    >
      Test #3
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Select))>
  <WithStyles(ForwardRef(FormHelperText))>
    Lorem ipsum dolor
  </WithStyles(ForwardRef(FormHelperText))>
</WithStyles(ForwardRef(FormControl))>
`;
